ocaml/xapi/xapi.ml

```diff
@@ -705,16 +705,21 @@ let set_stunnel_legacy_inv () =
 
 (* Consult database, in case inventory was out of date due to a DB change while
  * we were shut down. *)
-let set_stunnel_legacy_db ~__context () =
-  let localhost = Helpers.get_localhost ~__context in
-  let legacy = Db.Host.get_ssl_legacy ~__context ~self:localhost in
-  debug "Setting client stunnel legacy %b based on database lookup." legacy;
-  Stunnel.set_legacy_protocol_and_ciphersuites_allowed legacy;
-  let inv = Xapi_inventory.lookup Xapi_inventory._stunnel_legacy ~default:"true" in
-  let s = string_of_bool legacy in
-  if (s <> inv) then (
-    info "Host.ssl_legacy differs from inventory %s. Updating inventory to %s." Xapi_inventory._stunnel_legacy s;
-    Xapi_inventory.update Xapi_inventory._stunnel_legacy (s);
+let set_stunnel_legacy_db () =
+  Server_helpers.exec_with_new_task "Reconfig (from DB) for outgoing stunnels" (fun __context ->
+    let legacy_cur = Stunnel.is_legacy_protocol_and_ciphersuites_allowed () in
+    let localhost = Helpers.get_localhost ~__context in
+    let legacy_db = Db.Host.get_ssl_legacy ~__context ~self:localhost in
+    if legacy_db <> legacy_cur then begin
+      debug "Setting stunnel legacy %b based on database lookup." legacy_db;
+      Stunnel.set_legacy_protocol_and_ciphersuites_allowed legacy_db;
```

Comment:
We should do some refactoring here: everything (except the logging strings) in this if-block can be shared with `Xapi_host.set_ssl_legacy`, which should do the same things plus setting the database value.